src/shared/sam/debugger/awsSamDebugger.ts

```diff
@@ -25,18 +36,129 @@ export class AwsSamDebugConfigurationProvider implements vscode.DebugConfigurati
+        )
+    }
+
+    if (!AWS_SAM_DEBUG_TARGET_TYPES.includes(debugConfiguration.invokeTarget.target)) {
+        return localize(
+            'AWS.sam.debugger.invalidTarget',
+            'Debug Configuration has an unsupported target type. Supported types: {0}',
+            AWS_SAM_DEBUG_TARGET_TYPES.join(', ')
+        )
+    }
+}
+
+function templateDebugConfigValidation(
+    debugConfiguration: AwsSamDebuggerConfiguration,
+    cftRegistry: CloudFormationTemplateRegistry
+): string | undefined {
+    const templateTarget = (debugConfiguration.invokeTarget as any) as AwsSamDebuggerInvokeTargetTemplateFields
+
+    const template = cftRegistry.getRegisteredTemplate(templateTarget.samTemplatePath)
+
+    if (!template) {
+        return localize(
+            'AWS.sam.debugger.missingTemplate',
+            'Unable to find the Template file {0}',
+            templateTarget.samTemplatePath
+        )
+    }
+
+    const resources = template.template.Resources
+
+    if (!resources || !Object.keys(resources).includes(templateTarget.samTemplateResource)) {
+        return localize(
+            'AWS.sam.debugger.missingResource',
+            'Unable to find the Template Resource {0} in Template file {1}',
+            templateTarget.samTemplateResource,
+            templateTarget.samTemplatePath
+        )
+    }
+
+    const resource = resources[templateTarget.samTemplateResource]
+
+    // TODO: Validate against `AWS::Lambda::Function`?
+    if (resource?.Type !== CloudFormation.SERVERLESS_FUNCTION_TYPE) {
+        return localize(
+            'AWS.sam.debugger.resourceNotAFunction',
+            'Template Resource {0} in Template file {1} needs to be of type {2}',
+            templateTarget.samTemplateResource,
+            templateTarget.samTemplatePath,
+            CloudFormation.SERVERLESS_FUNCTION_TYPE
+        )
+    }
+
+    if (!resource?.Properties?.Runtime || !samLambdaRuntimes.has(resource?.Properties?.Runtime as string)) {
+        return localize(
+            'AWS.sam.debugger.unsupportedRuntime',
+            'Runtime for Template Resource {0} in Template file {1} is either undefined or unsupported.',
+            templateTarget.samTemplateResource,
+            templateTarget.samTemplatePath
+        )
+    }
+
+    const templateEnv = resource?.Properties.Environment
+    if (templateEnv?.Variables) {
+        const templateEnvVars = Object.keys(templateEnv.Variables)
+        const missingVars: string[] = []
+        if (debugConfiguration.lambda && debugConfiguration.lambda.environmentVariables) {
+            for (const key of Object.keys(debugConfiguration.lambda.environmentVariables)) {
+                if (!templateEnvVars.includes(key)) {
+                    missingVars.push(key)
+                }
+            }
+        }
+        if (missingVars.length > 0) {
+            // this check doesn't affect template validity.
+            vscode.window.showInformationMessage(
```

Comment:
Its weird to return validation strings but then message this warning string. It would be ideal for this function to not make any notifications at all.

Reply:
The problem is, this one isn't a failure message. I could do a `pair` style construct if you'd rather.

Reply:
As in returning something like `{ error?: string, warning?: string }` ?

Alternatively, split the warning check(s) to a separate method. The warning check method can make assumptions based on the error checks happening first.

I have no strong opinion either way.

Reply:
Went with a `tuple`, let me know what you think

Reply:
Please make it typed (the equivalent appears to be `{ validation: string, isError: boolean }`. The syntax for the tuple change looks difficult to read and work with.